File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/verbs/read_properties.py
from __future__ import annotations

import json
from pathlib import Path

from ..parser import MalformedFrontmatterError, MissingFrontmatterError
from ..prompt.read_properties import read_properties, read_properties_root


def _sort(v):
    if isinstance(v, list):
        return [_sort(x) for x in v]
    if isinstance(v, dict):
        return {k: _sort(v[k]) for k in sorted(v.keys())}
    return v


def read_properties_verb(args) -> tuple[int, str]:
    path = Path(args.path)
    try:
        target = path.resolve(strict=True)
    except (OSError, FileNotFoundError):
        return 2, f"no SKILL.md found under: {args.path}\n"

    if getattr(args, "root_mode", False):
        try:
            arr = read_properties_root(str(target))
        except (MissingFrontmatterError, MalformedFrontmatterError) as e:
            return 1, f"read-properties: {e}\n"
        except Exception as e:
            return 1, f"read-properties: {e}\n"
        if not arr:
            return 2, f"no SKILL.md found under: {args.path}\n"
        return 0, json.dumps([_sort(x) for x in arr], indent=2, ensure_ascii=False) + "\n"

    if not target.is_dir():
        return 2, f"no SKILL.md found under: {args.path}\n"
    if not (target / "SKILL.md").is_file():
        return 2, f"no SKILL.md found under: {args.path}\n"
    try:
        obj = read_properties(str(target))
    except (MissingFrontmatterError, MalformedFrontmatterError) as e:
        return 1, f"read-properties: {e}\n"
    except Exception as e:
        return 1, f"read-properties: {e}\n"
    return 0, json.dumps(_sort(obj), indent=2, ensure_ascii=False) + "\n"

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/cli/verbs/read-properties.mjs
import { statSync } from "node:fs";
import { resolve, join } from "node:path";
import {
  MalformedFrontmatterError,
  MissingFrontmatterError,
} from "../../parser.mjs";
import {
  readProperties,
  readPropertiesRoot,
} from "../../prompt/read_properties.mjs";

function parseArgs(args) {
  const opts = { path: null, rootMode: false };
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === "--root") {
      opts.rootMode = true;
      const v = args[++i];
      if (v && !v.startsWith("--")) opts.path = v;
    } else if (!a.startsWith("--") && opts.path == null) {
      opts.path = a;
    }
  }
  return opts;
}

function sortKeys(v) {
  if (Array.isArray(v)) return v.map(sortKeys);
  if (v && typeof v === "object") {
    const o = {};
    for (const k of Object.keys(v).sort()) o[k] = sortKeys(v[k]);
    return o;
  }
  return v;
}

export async function readPropertiesVerb(args) {
  const opts = parseArgs(args);
  if (!opts.path) {
    return {
      exitCode: 64,
      output:
        "usage: skills-ref read-properties <skill-dir> | --root <root-dir>\n",
    };
  }
  let target;
  try {
    target = resolve(opts.path);
    statSync(target);
  } catch {
    return { exitCode: 2, output: `no SKILL.md found under: ${opts.path}\n` };
  }

  if (opts.rootMode) {
    let arr;
    try {
      arr = readPropertiesRoot(target);
    } catch (e) {
      return { exitCode: 1, output: `read-properties: ${e.message}\n` };
    }
    if (arr.length === 0) {
      return { exitCode: 2, output: `no SKILL.md found under: ${opts.path}\n` };
    }
    return {
      exitCode: 0,
      output: JSON.stringify(arr.map(sortKeys), null, 2) + "\n",
    };
  }

  // single-skill mode: target must be a directory containing SKILL.md
  let st;
  try {
    st = statSync(target);
  } catch {
    return { exitCode: 2, output: `no SKILL.md found under: ${opts.path}\n` };
  }
  if (!st.isDirectory()) {
    return { exitCode: 2, output: `no SKILL.md found under: ${opts.path}\n` };
  }
  try {
    statSync(join(target, "SKILL.md"));
  } catch {
    return { exitCode: 2, output: `no SKILL.md found under: ${opts.path}\n` };
  }
  let obj;
  try {
    obj = readProperties(target);
  } catch (e) {
    if (
      e instanceof MissingFrontmatterError ||
      e instanceof MalformedFrontmatterError
    ) {
      return { exitCode: 1, output: `read-properties: ${e.message}\n` };
    }
    return { exitCode: 1, output: `read-properties: ${e.message}\n` };
  }
  return { exitCode: 0, output: JSON.stringify(sortKeys(obj), null, 2) + "\n" };
}
